### DISCUSSION TITLE
String Question

### DISCUSSION THREAD

**Discussion by ghost** [cc65/cc65#1446] [Q&A]
I am not 100% sure if CC65 handles strings in a non standard way. But I am running into issues with resetting a string that lists possible keyboard options. A command like  MYSTR = ""; or MYSTR = "Options:";  should reset the string back to a null state.  But I am getting "Option: One TwoOpenions: Two" when the function is called multiple times.

**Comment by mrdudz** [comment]
C does not "handle strings" at all. And it doesn't work like this in C

**Reply by ghost** [reply]
I might need to look at an alternative with just printing the list on the same line.

**Reply by Compyx** [reply]
You need to learn C. And using cc65 isn't the way to go about it, using C on a very restricted system like a 6502-based computer will only teach you 'bad' habits. Start with a system that supports proper C99 or so.

**Reply by polluks** [reply]
> Start with a system that supports proper C99 or so.

Well, we support proper C99 (beside floats). However a system with a source level debugger may help peteym5 or waiting for "db65" ;-)

**Reply by polluks** [reply]
> I might need to look at an alternative with just printing the list on the same line.

Do you know the important difference between `str=""` and `strcpy(str, "")`?

**Reply by Compyx** [reply]
I assume you know what I meant. An 8-bit target with all it's limitations and weird libs isn't ideal to learn C.

**Reply by polluks** [reply]
I'm not sure. Did you try to read the output of a compiler? 6502 is much more easier to understand than RISC for example.

**Reply by mrdudz** [reply]
> we support proper C99

Well well. Still a long way to go for that :) We support a reasonable subset :=)

**Comment by oliverschmidt** [comment]
Trying to be as disambiguous as possible: Don't expect to get any answers to this type of questions here! Neither as issue nor as discussion!

**Reply by polluks** [reply]
@peteym5 `MYSTR = "Options:"` who teaches you this kind of uppercase convention? 
It reminds me of the book "Russia For Dummies": We don't use these strange Russian letters, just the normal ones.
I was thinking immediately: Burn it, burn it! :-)

**Comment by greg-king5** [comment]
C isn't like BASIC.  In BASIC, strings are "simple" objects.  We can use them in almost the same way that we use numbers.  We can write:
```
first$ = "one"
second$ = "two"
third$ = first$ + second$
```
`third$` becomes a string that holds what you expect it to hold: `"onetwo"`.

But, C isn't that easy.  It doesn't do, automatically, the low-level steps that BASIC interpreters do.  You must write those steps yourself.  For example:
```c
char *first;
char *second;
char third[10];  // Must be at least 3 + 3 + 1 chars long, for this example.

first = "one";
second = "two";
strcpy(third, first);  // This is how you "reset" third.
strcat(third, second);
```
`third$` was created automatically by BASIC.  But in C, you must define `third[]` explicitly.  You must make it be as big as the longest string that you plan to build in it (remember that strings are stored with an extra byte).

In BASIC, `first$` and `second$` are strings.  But in C, `first` and `second` are pointers.  In the example, they point to string literals.  Those literals sit in a place that must not be changed.  The first arguments of `strcpy()` and `strcat()` point to places that will be changed by those functions.  Therefore, you must _not_ write code such as
```c
strcat(first, second);  // First argument points to a place that shouldn't be changed!
```

**Reply by polluks2** [reply]
", 3"? Looks like strncpy and strncat ;-)

**Reply by greg-king5** [reply]
Oops, you are right.  I better fix that fast. :wink:
[Although I typed `strcpy()`, my brain was thinking `memcpy()`.]

**Comment by ghost** [comment]
Thankyou very much. I just use a gotoxy() and just cputs() on the screen with the list of what I need to show. I did play around with STRCPY and STRCAT.

**Comment by ghost** [comment]
https://github.com/peteym5/Black-Jack

This is what I am working on and is the program that needs some string (or text data) manipulation. It is a Black Jack game I ported over from Fast Basic and using it as a learner project. It may be portable to other systems like the Commodore 64, 16, +4, Apple II. But the main thing is to get it working on the 8-bit.

**Comment by mrdudz** [comment]
briefly skimmed over it, some things you might want to change to make it actually portable:
- use cprintf instead of print. ONLY ever use the screen i/o functions from conio.h, do NOT mix with functions from stdio/stdlib
- use character literals instead of integer values when checking input, ie 'a' instead of 65
- get rid of PEEK and POKE as much as possible. not only this is very "non C", its also non portable. i'd suggest to replace as much as possible by library functions (eg for setting colors) and what remains can be moved into seperate functions, which you can then make target specific (perhaps a seperate file that gets linked per target)
good luck :)
